rxbindings/src/main/java/com/amplifyframework/rx/RxApiBinding.java

```diff
@@ -168,4 +173,83 @@
             CancelableBehaviors.StreamEmitter<String, T, ApiException> method) {
         return CancelableBehaviors.toObservable(method);
     }
+
+    /**
+     * A class that represents a subscription operation and exposes
+     * observables for consumers to listen to subscription data and
+     * status events.
+     * @param <T> The type representing the subscription data.
+     */
+    public static final class RxSubscriptionOperation<T> implements Cancelable {
+        private BehaviorSubject<ConnectionState> connectionStateSubject;
+        private Observable<T> subscriptionData;
+        private Cancelable amplifyOperation;
+        private String subscriptionId;
+
+        private Consumer<String> onConnected = new Consumer<String>() {
+            @Override
+            public void accept(@NonNull String subscriptionId) {
+                RxSubscriptionOperation.this.subscriptionId = subscriptionId;
+                connectionStateSubject.onNext(ConnectionState.CONNECTED);
+            }
+        };
+
+        RxSubscriptionOperation(CancelableBehaviors.StreamEmitter<String, T, ApiException> callbacks) {
+            connectionStateSubject = BehaviorSubject.create();
+            subscriptionData = Observable.create(emitter -> {
+                amplifyOperation = callbacks.streamTo(onConnected::accept,
+                                                      emitter::onNext,
+                                                      emitter::onError,
+                                                      emitter::onComplete);
+            });
+        }
+
+        /**
+         * Returns an {@link Observable} which consumers can use to
+         * retrieve data received by the subscription operation.
+         * @return Reference to the {@link Observable} with subscription data.
+         */
+        public Observable<T> observeSubscriptionData() {
+            return subscriptionData;
+        }
+
+        /**
+         * Once the subscription starts, this method returns
+         * the value of the subscriptionId.
+         * @return The value of the subscriptionId.
+         */
+        public String getSubscriptionId() {
```

Comment:
Is there any way to bundle this into the `ConnectionState` object?

Reply:
Good call on this one. Rather that doing it this way, I created a ConnectionStateEvent that has a connectionState and subscriptionId properties. We can expand on that down the road.